File: JS-Main/calc.js
const STAGE = {
	DEGREE: 1,
	MINUTE: 2,
	SECOND: 3,
	CARDINAL: 4,
};

function degree_decimal(degree, minute, second)
{
	return degree + (minute/60) + (second/3600);
}

function read_coordinate(rawdata)
{
	current_stage = STAGE.DEGREE;
	let degree = 0.0;
	let minute = 0.0;
	let second = 0.0;
	let south = false;

	for (let i = 0; i<rawdata.length; i++)
	{
		if (rawdata[i] === 'd') { current_stage = STAGE.MINUTE; continue; }
		if (rawdata[i] === '\'') { current_stage = STAGE.SECOND; continue; }
		if (rawdata[i] === '\"') { current_stage = STAGE.CARDINAL; continue; }
		switch (current_stage)
		{
			case STAGE.DEGREE:
				degree = degree*10 + (+rawdata[i]);
				break;
			case STAGE.MINUTE:
				minute = minute*10 + (+rawdata[i]);
				break;
			case STAGE.SECOND:
				second = second*10 + (+rawdata[i]);
				break;
			case STAGE.CARDINAL:
				if (rawdata[i] === 's' || rawdata[i] === 'S') { south = true;}
				break;
			default:
				break;
		}
	}

	let sum = degree_decimal(degree, minute, second);
	if (south === true) { sum *= -1; }

	return sum;
}

File: JS-Main/depricated/calc.lua
STAGE = {
	DEGREE = 1,
	MINUTE = 2,
	SECOND = 3,
	CARDINAL = 4,
}

function degree_decimal(degree, minute, second)
	return degree + math.floor(minute/60) + math.floor(second/3600)
end

function read_coordinate(input)
	current_stage = STAGE.DEGREE
	degree = 0.0
	minute = 0.0
	second = 0.0
	south = false

	for i = 1,input.len do
		if input[i] == 'd' or input[i] == 'D' then
			current_stage = STAGE.MINUTE
			goto continue
		end
		if input[i] == '\'' then
			current_stage = STAGE.SECOND
			goto continue
		end
		if input[i] == '\"' then
			current_stage = STAGE.CARDINAL
			goto continue
		end

		local case = {
			[STAGE.DEGREE] = function()
				degree = degree*10 + tonumber(input[i])
			end,
			[STAGE.MINUTE] = function()
				minute = minute*10 + tonumber(input[i])
			end,
			[STAGE.SECOND] = function()
				second = second*10 + tonumber(input[i])
			end,
			[STAGE.CARDINAL] = function()
				if input[i] == 's' or input[i] == "s" then
					south = true
				end
			end,
		}

		local switch = case[current_stage]
		if switch then
			switch()
		else
			print("an error has occured in the switch case")
		end

		::continue::
	end

	local sum = degree_decimal(degree, minute, second)
	if south then
		sum = sum * (-1)
	end

	return sum
end
